alerter.go

```diff
@@ -111,36 +111,38 @@ func ReportErrors(config Config, errors []string) []string {
 	return report
 }
 
-func included(logMessage string, includes []string) bool {
-	if len(includes) == 0 {
-		return true
-	}
-
-	for _, include := range includes {
-		matched, err := regexp.MatchString(include, logMessage)
-		if err != nil {
-			panic(err)
-		}
+type PatternMatcher struct {
+	includes []*regexp.Regexp
+	excludes []*regexp.Regexp
+}
 
-		if matched {
-			return true
+func NewPatternMatcher(includes []string, excludes []string) *PatternMatcher {
+	toRegexp := func(patterns []string) []*regexp.Regexp {
+		allRegexp := make([]*regexp.Regexp, 0, len(patterns))
+		for _, pattern := range patterns {
+			allRegexp = append(allRegexp, regexp.MustCompile(pattern))
 		}
+		return allRegexp
 	}
-	return false
+	return &PatternMatcher{includes: toRegexp(includes), excludes: toRegexp(excludes)}
 }
 
-func excluded(logMessage string, excludes []string) bool {
-	for _, exclude := range excludes {
-		matched, err := regexp.MatchString(exclude, logMessage)
-		if err != nil {
-			panic(err)
+func (matcher *PatternMatcher) FindAllMatch(s string) []string {
+	matches := make([]string, 0, 0)
+	for _, include := range matcher.includes {
+		for _, match := range include.FindAllString(s, -1) {
+			matches = append(matches, match)
 		}
+	}
 
-		if matched {
-			return true
+	for _, exclude := range matcher.excludes {
+		for i, match := range matches {
+			if len(exclude.FindAllString(match, -1)) != 0 {
+				matches = append(matches[:i], matches[i+1:]...)
```

Comment:
this is modifying the slice you are iterating over, and you are creating a new slice without the element at the current index, will the original iteration not go next to `i+1` which in the new array was `i+2`, skipping the old `i+1` ?
eg, 
includes "1.*"
excludes "11"
["0","1","11","11", "2"]
includes creates: ["1", "11", "11"]
exlcludes:
1) 0, "1",  exclude false, ["1", "11", "11"]
2) 1, "11", exclude true, ["1"] + ["11"] => ["1","11"]
3) 2, `?` => `panic: runtime error: slice bounds out of range`